tests/optimization/defer-structured-buffer-load.slang

```diff
@@ -0,0 +1,38 @@
+//TEST:SIMPLE(filecheck=CUDA): -target cuda -entry compute_main -stage compute
+//TEST:SIMPLE(filecheck=SPV): -target spirv
+
+// Test that we can defer loading big structured buffer elements.
```

Comment:
I don't have detailed comments on the various test cases, but I want to comment on how happy I am to see a large number of new test cases added with this work.

My largest concern when reading the code was that there can be a ton of subtlety in the logic to reason about aliasing, side effects, immutability, etc. and bugs in that logic can be extremely subtle to identify and track down (since even with an incorrect analysis, the compiler will often coincidentally end up generating code that works). Having a bunch of tests to hammer on this stuff is good.